src/Bicep.Core/Parsing/Parser.cs

```diff
@@ -1,370 +1,184 @@
+                        declarationsOrTokens.Add(newLine);
+                    }
+                }
+            }
+
+            var endOfFile = reader.Read();
+
+            return new ProgramSyntax(declarationsOrTokens, endOfFile, this.lexerDiagnostics);
+        }
+
+        public SyntaxBase Declaration() =>
+            this.WithRecovery(
+                () =>
+                {
+                    List<SyntaxBase> leadingNodes = new();
+
+                    // Parse decorators before the declaration.
+                    if (this.Check(TokenType.At))
                     {
-                        declarationsOrTokens.Add(newLine);
-                    }
-                }
-            }
-
-            var endOfFile = reader.Read();
-
-            return new ProgramSyntax(declarationsOrTokens, endOfFile, this.lexerDiagnostics);
-        }
-
-        public SyntaxBase Declaration() =>
-            this.WithRecovery(
-                () =>
-                {
-                    var current = reader.Peek();
-
-                    return current.Type switch
-                    {
-                        TokenType.Identifier => current.Text switch
-                        {
-                            LanguageConstants.TargetScopeKeyword => this.TargetScope(),
-                            LanguageConstants.ParameterKeyword => this.ParameterDeclaration(),
-                            LanguageConstants.VariableKeyword => this.VariableDeclaration(),
-                            LanguageConstants.ResourceKeyword => this.ResourceDeclaration(),
-                            LanguageConstants.OutputKeyword => this.OutputDeclaration(),
-                            LanguageConstants.ModuleKeyword => this.ModuleDeclaration(),
-                            _ => throw new ExpectedTokenException(current, b => b.UnrecognizedDeclaration()),
-                        },
-                        TokenType.NewLine => this.NewLine(),
-
-                        _ => throw new ExpectedTokenException(current, b => b.UnrecognizedDeclaration()),
-                    };
-                },
-                RecoveryFlags.None,
-                TokenType.NewLine);
-
-        private static RecoveryFlags GetSuppressionFlag(SyntaxBase precedingNode)
+                        while (true)
+                        {
+                            if (this.Check(TokenType.At))
+                            {
+                                leadingNodes.Add(this.Decorator());
+                                continue;
+                            }
+
+                            if (this.Check(TokenType.NewLine))
+                            {
+                                var newLine = this.WithRecoveryNullable(this.NewLineOrEof, RecoveryFlags.ConsumeTerminator, TokenType.NewLine);
+
+                                if (newLine != null)
+                                {
+                                    leadingNodes.Add(newLine);
+                                }
```

Comment:
Since you've already verified the next token is a newline, you can just do: `leadingNodes.Add(this.NewLine());` rather than invoking the recovery logic

Reply:
Good catch! Removed the recovery logic.